### PR TITLE
New headers first flow

### PR DESCRIPTION
**Pull request by someone235**

### CODE REVIEW

File: domain/consensus/processes/headersselectedtipmanager/headertipsmanager.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+func (h *headerTipsManager) AddHeaderTip(hash *externalapi.DomainHash) error {
+	hasSelectedTip, err := h.headersSelectedTipStore.Has(h.databaseContext)
+	if err != nil {
+		return err
+	}
+
+	if hasSelectedTip {
+		h.headersSelectedTipStore.Stage(hash)
```
Comment (elichai) [inline domain/consensus/processes/headersselectedtipmanager/headertipsmanager.go:36]: This is wrong.
You want to stage a new selectedTip only if it *doesn't* yet have any.
if it does you want to call `ChooseSelectedParent`

File: domain/consensus/processes/pruningmanager/pruningmanager.go [hunk 2]
```diff
@@ -233,6 +229,19 @@ func (pm *pruningManager) deleteBlock(blockHash *externalapi.DomainHash) (alread
...
 
+func (pm *pruningManager) CalculateIndependentPruningPoint(blockHash *externalapi.DomainHash) (*externalapi.DomainHash, error) {
+	ghostdagData, err := pm.ghostdagDataStore.Get(pm.databaseContext, blockHash)
+	if err != nil {
+		return nil, err
+	}
+
+	targetBlueScore := uint64(0)
+	if ghostdagData.BlueScore() > pm.pruningDepth {
+		targetBlueScore = ((ghostdagData.BlueScore() - pm.pruningDepth) / pm.finalityInterval) * pm.finalityInterval
```
Comment (elichai) [inline domain/consensus/processes/pruningmanager/pruningmanager.go:240]: This seem correct I think

File: domain/consensus/processes/blockvalidator/block_header_in_context.go [hunk 3]
```diff
@@ -60,7 +60,7 @@ func (v *blockValidator) ValidateHeaderInContext(blockHash *externalapi.DomainHa
 	return nil
 }
 
-func (v *blockValidator) isHeadersOnlyBlock(blockHash *externalapi.DomainHash) (bool, error) {
+func (v *blockValidator) hasValidatedHeader(blockHash *externalapi.DomainHash) (bool, error) {
```
Comment (svarogg) [inline domain/consensus/processes/blockvalidator/block_header_in_context.go:63]: rename to hasValidatedHeaderOnly

File: domain/consensus/processes/blockvalidator/proof_of_work.go [hunk 4]
```diff
@@ -101,7 +101,7 @@ func (v *blockValidator) checkProofOfWork(header *externalapi.DomainBlockHeader)
 func (v *blockValidator) checkParentsExist(blockHash *externalapi.DomainHash, header *externalapi.DomainBlockHeader) error {
 	missingParentHashes := []*externalapi.DomainHash{}
 
-	isFullBlock, err := v.blockStore.HasBlock(v.databaseContext, blockHash)
+	hashBlockBody, err := v.blockStore.HasBlock(v.databaseContext, blockHash)
```
Comment (svarogg) [inline domain/consensus/processes/blockvalidator/proof_of_work.go:104]: typo: has**h**BlockBody

File: domain/consensus/processes/headersselectedtipmanager/headertipsmanager.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	ghostdagManager         model.GHOSTDAGManager
+	headersSelectedTipStore model.HeaderSelectedTipStore
+}
+
+// New instantiates a new HeadersSelectedTipManager
+func New(databaseContext model.DBReader,
+	dagTopologyManager model.DAGTopologyManager,
+	ghostdagManager model.GHOSTDAGManager,
+	headersSelectedTipStore model.HeaderSelectedTipStore) model.HeadersSelectedTipManager {
+	return &headerTipsManager{
```
Comment (svarogg) [inline domain/consensus/processes/headersselectedtipmanager/headertipsmanager.go:21]: Please newline after multi-line func declaration

File: domain/consensus/consensus.go [hunk 6]
```diff
@@ -183,11 +184,38 @@ func (s *consensus) GetPruningPointUTXOSet(expectedPruningPointHash *externalapi
...
 	s.lock.Lock()
 	defer s.lock.Unlock()
 
-	return s.consensusStateManager.SetPruningPointUTXOSet(serializedUTXOSet)
+	selectedTip, err := s.headersSelectedTipStore.HeadersSelectedTip(s.databaseContext)
+	if err != nil {
+		return err
+	}
+
+	expectedNewPruningPointHash, err := s.pruningManager.CalculateIndependentPruningPoint(selectedTip)
```
Comment (svarogg) [inline domain/consensus/consensus.go:196]: Here we calculatePruningPoint according to selectedTip, while in another place we do according to virtual.
Wouldn't this create different pruning points whether you are IBDing or not?

File: domain/consensus/processes/pruningmanager/pruningmanager.go (hunk 2 above)
Comment (svarogg) [inline domain/consensus/processes/pruningmanager/pruningmanager.go:240]: Please write a comment explaining what exactly you are doing (I assume you want to round this to finalityIntervals, but this is far from trivial - so please comment)

**Review by svarogg — CHANGES_REQUESTED** [review]
The whole flow is lacking comments and logs.
This is a very complicated flow, many assumptions are hidden in code that is far away.

Therefore, please go over the whole flow, and add comments + logs at any spot that is not 100% trivial

File: app/protocol/flows/blockrelay/handle_relay_invs.go [hunk 7]
```diff
@@ -213,7 +213,11 @@ func (flow *handleRelayInvsFlow) processBlock(block *externalapi.DomainBlock) ([
 		}
 		log.Warnf("Rejected block %s from %s: %s", blockHash, flow.peer, err)
 
-		return nil, protocolerrors.Wrapf(true, err, "got invalid block %s from relay", blockHash)
+		// If we got in relay a block where one of its parents is missing a body it means that the peer that sent
```
[the remark this answers is not in the document]
Comment (someone235) [reply app/protocol/flows/blockrelay/handle_relay_invs.go:216]: @elichai Please approve this change

File: domain/consensus/consensus.go (hunk 6 above)
Comment (someone235) [reply domain/consensus/consensus.go:196]: In UpdatePruningPointByVirtual it doesn't update pruning point unless `if virtual.BlueScore() <= currentPBlueScore+pm.finalityInterval {`

**Review by svarogg — CHANGES_REQUESTED** [review]

**Review by svarogg — DISMISSED** [review]

**Comment by svarogg** [comment]
Rebuild after update in master

**Comment by svarogg** [comment]
Rebuild after update in master

**Review by svarogg — APPROVED** [review]

**Pull request merged**